pkg/agent/watcher.go

```diff
@@ -84,78 +86,47 @@ func (w *Watcher) Start() {
 			}
 		}
 	}()
-	err = watcher.Add(w.ConfigDir)
-	if err != nil {
-		log.Fatal(err)
-	}
-	log.Println("Watching", w.ConfigDir)
+	log.Println("Watching", w.configDir)
 	<-done
 }
 
-func (w *Watcher) ParseConfig(modelConfigs modelconfig.ModelConfigs) {
-	timeNow := time.Now()
+func (w *Watcher) parseConfig(modelConfigs modelconfig.ModelConfigs) {
 	for _, modelConfig := range modelConfigs {
-		modelName := modelConfig.Name
-		modelSpec := modelConfig.Spec
-		log.Println("Name:", modelName, "Spec:", modelSpec)
-		oldModel, ok := w.ModelTracker[modelName]
-		if !ok {
-			w.ModelTracker[modelName] = ModelWrapper{
-				ModelSpec:  &modelSpec,
-				Time:       timeNow,
-				Stale:      true,
-				Redownload: true,
-			}
-		} else {
-			isStale := true
-			reDownload := true
-			if oldModel.ModelSpec != nil {
-				isStale = !cmp.Equal(*oldModel.ModelSpec, modelSpec)
-				reDownload = !cmp.Equal(oldModel.ModelSpec.StorageURI, modelSpec.StorageURI)
-				log.Println("same", !isStale, *oldModel.ModelSpec, modelSpec)
-			}
-			// Need to store new time, TODO: maybe worth to have seperate map?
-			w.ModelTracker[modelName] = ModelWrapper{
-				ModelSpec:  &modelSpec,
-				Time:       timeNow,
-				Stale:      isStale,
-				Redownload: reDownload,
-			}
+		name, spec := modelConfig.Name, modelConfig.Spec
+		existing, exists := w.modelTracker[name]
+		if !exists {
+			// New - add
+			w.modelTracker[name] = modelWrapper{Spec: &spec}
+			w.modelAdded(name, &spec)
+		} else if !cmp.Equal(spec, *existing.Spec) {
+			existing.Spec, existing.stale = &spec, false
+			// Changed - replace
+			w.modelRemoved(name)
+			w.modelAdded(name, &spec)
```

Comment:
Should this be just Added/Reload ? Remove then Add triggers unload and load which would cause some downtime of the model on the model server side?

Reply:
My thinking was that if TrainedModels are immutable (as I think was agreed), this case is unlikely to occur since it would imply the TrainedModel CR itself had been deleted and another with the same name subsequently created. In that case continuity would not be expected anyhow.

Reply:
As I said above, I think that we should handle updates to the TrainedModel.